src/blas/backends/rocblas/rocblas_level1.cpp

```diff
@@ -0,0 +1,1758 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+#include "rocblas_helper.hpp"
+#include "rocblas_task.hpp"
+
+#include "oneapi/mkl/exceptions.hpp"
+#include "oneapi/mkl/blas/detail/rocblas/onemkl_blas_rocblas.hpp"
+
+
+namespace oneapi {
+namespace mkl {
+namespace blas {
+namespace rocblas {
+namespace column_major {
+
+// Buffer APIs
+
+// Level 1
+template <typename Func, typename T1, typename T2>
+inline void asum(Func func, cl::sycl::queue &queue, int64_t n, cl::sycl::buffer<T1, 1> &x,
+                 const int64_t incx, cl::sycl::buffer<T2, 1> &result) {
+    using cuDataType1 = typename CudaEquivalentType<T1>::Type;
+    using cuDataType2 = typename CudaEquivalentType<T2>::Type;
+    overflow_check(n, incx);
+
+    queue.submit([&](cl::sycl::handler &cgh) {
+        auto x_acc = x.template get_access<cl::sycl::access::mode::read>(cgh);
+        auto res_acc = result.template get_access<cl::sycl::access::mode::write>(cgh);
+        onemkl_rocblas_host_task(cgh, queue,[=](RocblasScopedContextHandler& sc) {
+            auto handle = sc.get_handle(queue);
+            // By default the pointer mode is the rocblas_pointer_mode_host
+            // when the data is on buffer, it must be set to
+            // rocblas_set_pointer_mode mode otherwise it causes the segmentation
+            // fault. When it is set to device it is users responsibility to
+            // synchronise as the function is completely asynchronous.
+            rocblas_set_pointer_mode(handle, rocblas_pointer_mode_device);
+            auto x_ = sc.get_mem<cuDataType1 *>(x_acc);
+            auto res_ = sc.get_mem<cuDataType2 *>(res_acc);
+            rocblas_status err;
+            // ASUM does not support negative index
+            ROCBLAS_ERROR_FUNC(func, err, handle, n, x_, std::abs(incx), res_);
+            // Higher level BLAS functions expect rocblas_pointer_mode_host
+            // to be set, therfore we need to reset this to the default value
+            // in order to avoid CUDA_ERROR_ILLEGAL_ADRESS errors
```

Comment:
Is `CUDA_ERROR_ILLEGAL_ADDRESS` what you mean here?

Reply:
This referes to the issue described in #92 and fixed in #94. `CUDA_ERROR_ILLEGAL_ADRESES` will need to be changed to the rocm eqvivalent.

Reply:
please update as needed.